llvm/lib/Target/AArch64/AArch64FrameLowering.cpp

```diff
@@ -3201,11 +3196,27 @@ bool AArch64FrameLowering::restoreCalleeSavedRegisters(
     return true;
   }
 
+  // For performance reasons restore SVE register in increasing order
+  auto PPRBegin =
+      std::find_if(RegPairs.begin(), RegPairs.end(), [](const RegPairInfo &c) {
+        return c.Type == RegPairInfo::PPR;
+      });
+  auto PPREnd = std::find_if(
+      RegPairs.rbegin(), RegPairs.rend(),
+      [](const RegPairInfo &c) { return c.Type == RegPairInfo::PPR; });
+  std::reverse(PPRBegin, PPREnd.base());
+  auto ZPRBegin =
+      std::find_if(RegPairs.begin(), RegPairs.end(), [](const RegPairInfo &c) {
+        return c.Type == RegPairInfo::ZPR;
+      });
+  auto ZPREnd = std::find_if(
+      RegPairs.rbegin(), RegPairs.rend(),
+      [](const RegPairInfo &c) { return c.Type == RegPairInfo::ZPR; });
+  std::reverse(ZPRBegin, ZPREnd.base());
+
   if (ReverseCSRRestoreSeq) {
```

Comment:
If `ReverseCSRRestoreSeq == true`, then it always reverses all RegPairs. That means that if RegPairs has `{ ..., P4, P5, ... Z8, Z9, .. }` that if `ReverseCSRRestoreSeq == true`, that the Z registers will be restored first, which I thought is what you want to avoid (reading the commit messsage in the PR).

That said, I think this option is never really used other than in some of the tests. Which makes me wonder, can we remove it? @francisvm

Reply:
We can remove it! I added it for experiments at first and it didn't turn out useful in the end.